src/languages/rust.ts

```diff
@@ -0,0 +1,175 @@
+import { SyntaxNode } from "web-tree-sitter";
+import { SimpleScopeTypeType } from "../typings/targetDescriptor.types";
+import { NodeMatcherAlternative } from "../typings/Types";
+import { patternFinder } from "../util/nodeFinders";
+import {
+  ancestorChainNodeMatcher,
+  argumentMatcher,
+  cascadingMatcher,
+  createPatternMatchers,
+  leadingMatcher,
+  matcher,
+  patternMatcher,
+  trailingMatcher,
+} from "../util/nodeMatchers";
+
+// Generated by the following command:
+// `curl https://raw.githubusercontent.com/tree-sitter/tree-sitter-rust/36ae187ed6dd3803a8a89dbb54f3124c8ee74662/src/node-types.STATEMENT_TYPES | jq '[.[] | select(.type == "_declaration_statement") | .subtypes[].type, "expression_statement"]'`
+const STATEMENT_TYPES = [
+  "associated_type",
+  "attribute_item",
+  "const_item",
+  "empty_statement",
+  "enum_item",
+  "extern_crate_declaration",
+  "foreign_mod_item",
+  "impl_item",
+  "inner_attribute_item",
+  "let_declaration",
+  "macro_definition",
+  "macro_invocation",
+  "function_item",
+  "function_signature_item",
+  "mod_item",
+  "static_item",
+  "struct_item",
+  "trait_item",
+  "type_item",
+  "union_item",
+  "use_declaration",
+  "expression_statement",
+];
+
+/**
+ * Scope types allowed to be parents of a statement
+ */
+const STATEMENT_PARENT_TYPES = ["source_file", "block", "declaration_list"];
+
+/**
+ * Returns "impl_item[type]" node higher in the chain
+ * @param node The node which we will start our search from
+ * @returns node or null
+ */
+function traitTypeMatcher(node: SyntaxNode) {
+  var parentNode = node.parent;
+  // Try find parentNode which type is "impl_item"
+  while (true) {
```

Comment:
you don't need to loop here.  All of the marchers are called repeatedly as part of an algorithm that walks up the ancestors chain.  Just check that the node looks like you want it to, and return it if so

Reply:
I don’t think It’ll work. I want to match if I’m at the correct subtree. Please check.

Reply:
Your function is called as part of https://github.com/cursorless-dev/cursorless/blob/main/src/processTargets/modifiers/scopeTypeStages/ContainingSyntaxScopeStage.ts#L96-L110

Reply:
The code looks unchanged to me, so I'm not sure what to check?

Reply:
Discord?